Foreground Tissue Mask

drbeh opened the pull request:
Fixes #4380 
Prerequisite for #4014 

### Description
This PR implements `ForegroundMask` transform that creates a foreground mask with various thresholding. For digital pathology use cases, this transform is helpful to create sliding-window patches that covers only the tissue.

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

Nic-Ma commented on monai/transforms/intensity/array.py:
```diff
@@ -2161,3 +2157,97 @@ def __call__(self, img: torch.Tensor) -> torch.Tensor:
+            It also can be a callable that receives each dimension of the image and calculate the threshold,
+            or a string the defines such callable from `skimage.filter.threshold_xxxx`.
+            Moreover, a dictionary can be passed that defines such thresholds for each channel, like
+            {"R": 100, "G": "otsu", "B": skimage.filter.threshold_mean}
```
Could you please help add the reference link of `skimage.filter.threshold_XXX` here?

Thanks.

Nic-Ma commented on monai/transforms/intensity/array.py:
```diff
@@ -2161,3 +2157,97 @@ def __call__(self, img: torch.Tensor) -> torch.Tensor:
+    Args:
+        threshold: an int or a float number that defines the threshold that values less than that are foreground.
+            It also can be a callable that receives each dimension of the image and calculate the threshold,
+            or a string the defines such callable from `skimage.filter.threshold_xxxx`.
```
typo: `the` -> `that`.

Thanks.

Nic-Ma commented on monai/transforms/intensity/array.py:
```diff
@@ -2161,3 +2157,97 @@ def __call__(self, img: torch.Tensor) -> torch.Tensor:
+            rgb_foreground = np.zeros_like(img_rgb[:1])
+            for img, mode in zip(img_rgb, "RGB"):
+                threshold = self._get_threshold(img, mode)
+                if threshold:
```
Should we allow `threshold==None` here? I think maybe we should raise an exception or warning?

Thanks.

drbeh replied:
Now it raises an exception if the `threshold=None` or wrongly set.

Nic-Ma replied:
May I know where it raises an error now?
I mean currently, this function can return None: `_get_threshold`.

Thanks.

Nic-Ma replied:
OK, I see this part. If no None case in `_get_threshold`, seems we can remove this `if threshold:` line at line 2242 and line 2250?

Thanks.

Nic-Ma commented on monai/transforms/intensity/array.py:
```diff
@@ -2161,3 +2157,97 @@ def __call__(self, img: torch.Tensor) -> torch.Tensor:
+        if foregrounds:
+            mask = np.stack(foregrounds).all(axis=0)
+        else:
+            mask = np.zeros_like(img_rgb[:1])
```
Could you please help describe the `no foregrounds` case in the doc-string of this class?

Thanks.

drbeh replied:
I removed this cases as it tied to having no threshold.

Nic-Ma commented on monai/transforms/intensity/dictionary.py:
```diff
@@ -1654,6 +1658,49 @@ def __call__(self, data: Mapping[Hashable, NdarrayOrTensor]) -> Dict[Hashable, N
+        hsv_threshold: similar to threshold but HSV color space ("H", "S", and "V").
+            Unlike RBG, in HSV, value greater than `hsv_threshold` are considered foreground.
+        invert: invert the intensity range of the input image, so that the dtype maximum is now the dtype minimum,
+            and vice-versa.
```
missing doc-string for arg `new_key_prefix`.

Thanks.

Nic-Ma reviewed (commented):
Thanks for the new feature, it overall looks good to me, put some minor comments inline.

Nic-Ma reviewed (commented):
Thanks for the quick update, it overall looks good to me.
Put some minor comments inline.
As it's a typical use case for 2D image, could you please help add the sample image in the doc? Following: https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/utils_create_transform_ims.py

Thanks.

ericspod commented on monai/transforms/intensity/array.py:
```diff
@@ -2161,3 +2157,99 @@ def __call__(self, img: torch.Tensor) -> torch.Tensor:
         return img
+
+
+class ForegroundMask(Transform):
```
We should be changing the deformation transforms to have one class for 2D and 3D but regardless if we have a transform that is definitely only 2D it should say so in the name. It'll have to be something like `Foreground2DMask` since the name can't start with a number.

Nic-Ma replied:
`Foreground2DMask` looks good to me.

Thanks.

drbeh replied:
Thanks for your input. It should work for any dimensional input array. I don't know why did you get the impression that it works with only 2D. To make sure, I added a test case to include a 3D image.

Nic-Ma replied:
If it works for 3D, that's cool! Please add the 3D test.

Thanks.

drbeh commented:
> Thanks for the quick update, it overall looks good to me. Put some minor comments inline. As it's a typical use case for 2D image and in the core transforms, could you please help add the sample image in the doc? Following: https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/utils_create_transform_ims.py
> 
> Thanks.

@Nic-Ma, I think it has the same blocker that has not been resolved: https://github.com/Project-MONAI/MONAI/issues/4161

Nic-Ma commented:
Thanks for the quick update.
Looks good to me except for the missing sample image in doc.
@rijobro Could you please help guide @drbeh the basic steps to add a sample transform image?

Thanks in advance.

Nic-Ma commented:
> 

Hi @drbeh ,

I feel this transform is different from #4161 as this transform seems like a general-purpose intensity transform for 2D/3D images?
@ericspod @rijobro Please feel free to correct me if I misunderstood it.

Thanks.

drbeh commented:
HI @Nic-Ma, alright, let me see if I can create an image with the existing data for this transform.

drbeh commented:
@Nic-Ma, I've added the images for this transform: https://github.com/Project-MONAI/DocImages/pull/12, and updated this PR. Can you check they are fine?

Nic-Ma reviewed (approved):
Thanks for the quick update!
Looks good to me.

Thanks.

Nic-Ma commented:
/black

Nic-Ma commented:
/build
